### PR TITLE
Tamagotchi 03

### PR DESCRIPTION
**Pull request by Heriels-Sun**
This is my Tamagotchi - 03 Homeworl 😄

Here is the smart contract Vara Testnet link, Thanks! ✌️

[Tamagotchi - 03](https://idea.gear-tech.io/programs/0xff608024851d227ef7981256fedfbf7bfdbb639c374bb5bb564733649a5896e8?node=wss%3A%2F%2Ftestnet.vara-network.io)

### CODE REVIEW

File: contracts/03-tamagotchi-nft/io/src/lib.rs [hunk 1]
```diff
@@ -1,40 +1,87 @@
...
+use scale_info::TypeInfo;
 
 #[derive(Default, Encode, Decode, TypeInfo)]
 #[codec(crate = gstd::codec)]
 #[scale_info(crate = gstd::scale_info)]
 pub struct Tamagotchi {
-    // TODO: 0️⃣ Copy fields from previous lesson and push changes to the master branch
-    // TODO: 1️⃣ Add new fields
+    pub name: String, 
+    pub age: u64,
```
Comment (shamilsan) [inline contracts/03-tamagotchi-nft/io/src/lib.rs:13]:
```suggestion
    pub date_of_birth: u64,
```

File: contracts/03-tamagotchi-nft/io/src/lib.rs [hunk 2]
```diff
@@ -1,40 +1,87 @@
...
+    type State = Out<Tamagotchi>;
     type Reply = ();
     type Others = ();
     type Signal = ();
 }
+
+impl Tamagotchi {
+
+    pub fn calculate_hunger(&self) -> u64 {
+        HUNGER_PER_BLOCK * ((exec::block_timestamp() - self.fed_block) / 1000)
```
Comment (shamilsan) [inline contracts/03-tamagotchi-nft/io/src/lib.rs:66]: Use `exec::block_height()` instead of `exec::block_timestamp()`
```suggestion
        HUNGER_PER_BLOCK * (exec::block_height() - self.fed_block)
```

File: contracts/03-tamagotchi-nft/src/lib.rs [hunk 3]
```diff
@@ -1,19 +1,101 @@
...
 
-#[allow(unused_imports)]
-use gstd::prelude::*;
+static mut TAMAGOTCHI: Option<Tamagotchi> = None;
 
 #[no_mangle]
 extern fn init() {
-    // TODO: 0️⃣ Copy the `init` function from the previous lesson and push changes to the master branch
+    let name: String = msg::load().expect("Failed to decode Tamagotchi name");
+    let current_time_block = exec::block_timestamp();
```
Comment (shamilsan) [inline contracts/03-tamagotchi-nft/src/lib.rs:10]:
```suggestion
    let current_block = exec::block_height();
```

File: contracts/03-tamagotchi-nft/src/lib.rs [hunk 4]
```diff
@@ -1,19 +1,101 @@
...
+            msg::reply(TmgEvent::Name(tmg.name.clone()), 0)
+                .expect("Error in a reply `TmgEvent::Name`");
+        }
+        TmgAction::Age => {
+            let age = exec::block_timestamp() - tmg.age;
+            msg::reply(TmgEvent::Age(age), 0)
+                .expect("Error in a reply `TmgEvent::Age`");
+        }
+        TmgAction::Feed => {
+            tmg.fed += FILL_PER_FEED - tmg.calculate_hunger();
```
Comment (shamilsan) [inline contracts/03-tamagotchi-nft/src/lib.rs:46]: This subtraction can cause underflow if `tmg.calculate_hunger()` is greater than `FILL_PER_FEED`.

File: contracts/03-tamagotchi-nft/src/lib.rs [hunk 5]
```diff
@@ -1,19 +1,101 @@
...
+            tmg.fed_block = exec::block_timestamp();
```
Comment (shamilsan) [inline contracts/03-tamagotchi-nft/src/lib.rs:47]:
```suggestion
            tmg.fed_block = exec::block_height();
```

File: contracts/03-tamagotchi-nft/src/lib.rs [hunk 6]
```diff
@@ -1,19 +1,101 @@
...
+            tmg.fed = if tmg.fed > MAX_VALUE {
+                MAX_VALUE
+            } else {
+                tmg.fed
+            };
```
Comment (shamilsan) [inline contracts/03-tamagotchi-nft/src/lib.rs:52]: Might be shorter.
```suggestion
            tmg.fed = tmg.fed.min(MAX_VALUE);
```

File: contracts/03-tamagotchi-nft/src/lib.rs [hunk 7]
```diff
@@ -1,19 +1,101 @@
...
+        msg::reply(TmgEvent::Slept, 0).expect("Error in a reply `TmgEvent::Slept`");
+        },
+        TmgAction::Transfer(new_actor) => {
+            if source == tmg.owner || source == tmg.approved_account.unwrap() {
```
Comment (shamilsan) [inline contracts/03-tamagotchi-nft/src/lib.rs:78]: `unwrap` is a bad idea here as `approved_account` may be `None`

File: contracts/03-tamagotchi-nft/src/lib.rs [hunk 8]
```diff
@@ -1,19 +1,101 @@
...
+                tmg.owner = new_actor;
+            }
+            msg::reply(TmgEvent::Transferred(tmg.owner), 0).expect("Error in a reply `TmgEvent::Transferred`");
```
Comment (shamilsan) [inline contracts/03-tamagotchi-nft/src/lib.rs:81]: No need to reply with `Transferred` in all cases (even if no transfer occurred).

File: contracts/03-tamagotchi-nft/src/lib.rs [hunk 9]
```diff
@@ -1,19 +1,101 @@
...
+        },
+        TmgAction::Approve(new_actor) => {
+            if source == tmg.owner {
+                tmg.approved_account = Some(new_actor);
+            }
+            msg::reply(TmgEvent::Approved(tmg.approved_account.unwrap()), 0).expect("Error in a reply `TmgEvent::Approved`");
+        },
+        TmgAction::RevokeApproval => {
+            tmg.approved_account = None;
```
Comment (shamilsan) [inline contracts/03-tamagotchi-nft/src/lib.rs:90]: Only the owner (and maybe the approved account) can revoke approval.

File: contracts/03-tamagotchi-nft/tests/owning.rs [hunk 10]
```diff
@@ -1,12 +1,80 @@
-use gtest::{Program, System};
+use gtest::{Program, System, Log};
+use tamagotchi_io::*;
 
-// TODO: 0️⃣ Copy tests from the previous lesson and push changes to the master branch
+#[test]
+fn smoke_test_01() {
+    let sys = System::new();
+    sys.init_logger();
+    let _program = Program::current(&sys);
```
Comment (shamilsan) [inline contracts/03-tamagotchi-nft/tests/owning.rs:8]:
```suggestion
    let program = Program::current(&sys);
```

File: contracts/03-tamagotchi-nft/tests/owning.rs [hunk 11]
```diff
@@ -1,12 +1,80 @@
...
+
+    _program.send(2, String::from("Sofia"));
+
+    let _res = _program.send(2, TmgAction::Approve(3.into()));
+    let expected_log = Log::builder()
+    .dest(2)
+    .payload(TmgEvent::Approved(3.into()));
+    assert!(_res.contains(&expected_log));
+
+    let _res = _program.send(2, TmgAction::Transfer(3.into()));
```
Comment (shamilsan) [inline contracts/03-tamagotchi-nft/tests/owning.rs:68]: It would also be great to test negative scenarios (e.g., transfer by the not-approved account).
